src/FsAutoComplete.Core/AbstractClassStubGenerator.fs

```diff
@@ -84,77 +98,62 @@ let private tryFindAbstractClassExprInParsedInput
-  | AbstractClassData.ObjExpr(_, bindings, _) -> List.choose (|MemberNameAndRange|_|) bindings
+  | AbstractClassData.ObjExpr(bindings = bindings) -> List.choose (|MemberNamePlusRangeAndKeywordRange|_|) bindings
 
 /// Try to find the start column, so we know what the base indentation should be
 let inferStartColumn
-  (codeGenServer: ICodeGenerationService)
-  (pos: Position)
-  (doc: Document)
-  (lines: ISourceText)
-  (lineStr: string)
   (abstractClassData: AbstractClassData)
+  (memberNamesAndRanges: (_ * _ * Range) list)
   (indentSize: int)
   =
-  async {
-    match getMemberNameAndRanges abstractClassData with
-    | (_, range) :: _ -> return getLineIdent (lines.GetLineString(range.StartLine - 1))
-    | [] ->
-      match abstractClassData with
-      | AbstractClassData.ExplicitImpl _ ->
-        // 'interface ISomething with' is often in a new line, we use the indentation of that line
-        return getLineIdent lineStr + indentSize
-      | AbstractClassData.ObjExpr(_, _, newExprRange) ->
-        match! codeGenServer.TokenizeLine(doc.FullName, pos.Line) with
-        | Some tokens ->
-          return
-            tokens
-            |> List.tryPick (fun (t: FSharpTokenInfo) ->
-              if t.CharClass = FSharpTokenCharKind.Keyword && t.TokenName = "NEW" then
-                // We round to nearest so the generated code will align on the indentation guides
-                findGreaterMultiple (t.LeftColumn + indentSize) indentSize |> Some
-              else
-                None)
-            // There is no reference point, we indent the content at the start column of the interface
-            |> Option.defaultValue newExprRange.StartColumn
-        | None -> return newExprRange.StartColumn
-  }
+  match memberNamesAndRanges with
+  | (_, _, leadingKeywordRange) :: _ ->
+    // if we have any members, then we can use the start of the leading keyword to give us the indent correctly
+    leadingKeywordRange.StartColumn
+  | [] ->
+    match abstractClassData with
+    | AbstractClassData.ExplicitImpl(inheritExpressionRange = inheritRange) ->
+      // 'interface ISomething with' is often in a new line, we use the indentation of that line
+      inheritRange.StartColumn
+    | AbstractClassData.ObjExpr(newExpression = newExpr; withKeyword = withKeyword; bindings = bindings) ->
+      // two cases here to consider:
+      // * has a with keyword on same line as newExpr
+      match withKeyword with
+      | None ->
+        // if no withKeyword, then we add an indent to the start of the new Expression to get our final indent
+        newExpr.StartColumn + indentSize
+      | Some keyword ->
+        // if we have a keyword, if it's on the same line then we can do the same
+        if keyword.StartLine = newExpr.StartLine then
+          newExpr.StartColumn + indentSize
+        else if keyword.StartColumn = newExpr.StartColumn then
+          keyword.StartColumn + indentSize
+        else
+          keyword.StartColumn
 
```

Comment:
massive cleanup here thanks to AST additions - we can do all of these checks with absolute precision now. And synchronously!